File: daisy1-graalvm/src/main/java/io/openenterprise/graalvm/polyglot/PooledContextDelegateFactory.java
package io.openenterprise.graalvm.polyglot;

import jakarta.annotation.Nonnull;
import org.apache.commons.pool2.BasePooledObjectFactory;
import org.apache.commons.pool2.PooledObject;
import org.apache.commons.pool2.impl.DefaultPooledObject;

public class PooledContextDelegateFactory extends BasePooledObjectFactory<ContextDelegate> {

    public PooledContextDelegateFactory(@Nonnull ContextBuilder contextBuilder) {
        this.contextBuilder = contextBuilder;
    }

    protected ContextBuilder contextBuilder;

    @Override
    public void activateObject(@Nonnull PooledObject<ContextDelegate> pooledObject) throws Exception {
        pooledObject.getObject().enter();
    }

    @Override
    public ContextDelegate create() throws Exception {
        return new ContextDelegateImpl(contextBuilder.createContext(), contextBuilder.getSupportedLanguage());
    }

    @Override
    public void passivateObject(PooledObject<ContextDelegate> pooledObject) throws Exception {
        pooledObject.getObject().close();
    }

    @Override
    public PooledObject<ContextDelegate> wrap(@Nonnull ContextDelegate contextDelegate) {
        return new DefaultPooledObject<>(contextDelegate);
    }
}

File: daisy1-graalvm/src/main/kotlin/io/openenterprise/graalvm/polyglot/PooledContextDelegateFactory.kt
package io.openenterprise.graalvm.polyglot
import org.apache.commons.pool2.BasePooledObjectFactory
import org.apache.commons.pool2.PooledObject
import org.apache.commons.pool2.impl.DefaultPooledObject

class PooledContextDelegateFactory(val contextBuilder: ContextBuilder): BasePooledObjectFactory<ContextDelegate>() {

    override fun activateObject(p: PooledObject<ContextDelegate>) {
       p.`object`.enter()
    }

    override fun create(): ContextDelegateImpl {
        return ContextDelegateImpl(contextBuilder.createContext(), contextBuilder.supportedLanguage)
    }

    override fun wrap(p0: ContextDelegate): PooledObject<ContextDelegate> {
        return DefaultPooledObject(p0)
    }
}